tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/cleaning_tests_base.py

```diff
@@ -0,0 +1,229 @@
+
+
+class BaseTest:
+
+    class CleaningRulesTestBase(unittest.TestCase):
+        """
+        A base class for cleaning rules integration tests.
+
+        This class implements setUp, setUpClass, tearDown, and tearDownClass
+        methods that can be inherited and used by all extending classes.  This
+        allows developers to focus more fully on creating a working test rather
+        than setting up or tearing down their test environment.
+
+        The base class ensures test and sandbox datasets are created.  It also
+        creates empty test tables with correct schemas to further validate the
+        cleaning rule sql.  The base class ensures the test tables are reloaded
+        with the sample data between tests.  It also ensures tables are deleted
+        when the test is finished.
+        """
+
+        # this should always be for a test project.  a check is implemented to
+        # make sure this is true
+        project_id = ''
+        # a list of fully qualified table names the cleaning rule is targeting.
+        # fq = {project_id}.{dataset_id}.{table_name}.  required for
+        # test setup and cleanup.
+        fq_table_names = []
+        # a list of fully qualified sandbox tables the rule may create.
+        # required for cleanup
+        fq_sandbox_table_names = []
+        # a list of dictionaries where each dictionary defines table
+        # information for one table upon rule execution, such as the
+        # 'name', 'fq_table_name', 'fq_sanbox_table_name', 'loaded_ids',
+        # 'sandboxed_ids', and 'cleaned_ids'
+        tables_and_counts = []
+        # a list of load statements to execute before each test.  This can be
+        # defined in the extending class's 'setUp(self)' function.
+        sql_load_statements = []
+        # The query class that is being executed.
+        query_class = None
+
+        @classmethod
+        def setUpClass(cls):
+            # get the test project
+            if 'test' not in cls.project_id:
+                raise RuntimeError(
+                    'This should only be run on a test environment')
+
+            # get or create datasets, most cleaning rules assume the datasets exist
+            required_datasets = []
+            for table_info in cls.tables_and_counts:
+                dataset_id = table_info.get('fq_table_name', '').split('.')[1]
+                sandbox_id = table_info.get('fq_sandbox_table_name',
+                                            '').split('.')[1]
+                required_datasets.extend([dataset_id, sandbox_id])
+
+            for dataset_id in required_datasets:
+                bq.get_or_create_dataset(cls.project_id, dataset_id)
+
+            # create empty table in test environment
+            cls.client = bq.get_client(cls.project_id)
+
+            # the base class will try to determine this from the list of
+            # tables_and_counts if the list is empty
+            if not cls.fq_table_names:
```

Comment:
This is probably for a future iteration, but when I was writing an integration test, I found the lack of a concept and concept ancestor table somewhat problematic. Would it make sense to have those tables in particular be static, or have some default/seed content? In theory, there's no reason to avoid using the real concept* tables in the integration tests.

Reply:
This is a very good idea for a future iteration.  This might be something to store in a test bucket without a data retention policy that wipes it every 30 days.